keras_hub/src/models/d_fine/d_fine_utils.py

```diff
@@ -0,0 +1,530 @@
+        if method == "default":
+            sampling_value_l_ = grid_sample(
+                data=value_l_,
+                grid=sampling_grid_l_,
+                align_corners=False,
+                height=height,
+                width=width,
+            )
+        elif method == "discrete":
+            scale_factors = keras.ops.cast(
+                keras.ops.array([width, height]),
+                dtype=sampling_grid_l_.dtype,
+            )
+            sampling_coord_float = sampling_grid_l_ * scale_factors
+            _sampling_coord_x_int = keras.ops.cast(
+                keras.ops.floor(sampling_coord_float[..., 0] + 0.5), "int32"
+            )
+            _sampling_coord_y_int = keras.ops.cast(
+                keras.ops.floor(sampling_coord_float[..., 1] + 0.5), "int32"
+            )
+            clamped_coord_x = keras.ops.clip(
+                _sampling_coord_x_int, 0, width - 1
+            )
+            clamped_coord_y = keras.ops.clip(
+                _sampling_coord_y_int, 0, height - 1
+            )
+            sampling_coord_stacked = keras.ops.stack(
+                [clamped_coord_x, clamped_coord_y], axis=-1
+            )
+            B_prime = batch_size * num_heads
+            Q_dim = num_queries
+            P_level = num_points_list[level_id]
+            sampling_coord = keras.ops.reshape(
+                sampling_coord_stacked, (B_prime, Q_dim * P_level, 2)
+            )
+            value_l_permuted = keras.ops.transpose(value_l_, (0, 2, 3, 1))
+            y_coords_for_gather = sampling_coord[
+                ..., 1
+            ]  # (B_prime, Q_dim * P_level)
+            x_coords_for_gather = sampling_coord[
+                ..., 0
+            ]  # (B_prime, Q_dim * P_level)
+            indices = y_coords_for_gather * width + x_coords_for_gather
+            indices = keras.ops.expand_dims(indices, axis=-1)
+            value_l_flat = keras.ops.reshape(
+                value_l_permuted, (B_prime, height * width, hidden_dim)
+            )
+            gathered_values = keras.ops.take_along_axis(
+                value_l_flat, indices, axis=1
+            )
+            permuted_gathered_values = keras.ops.transpose(
+                gathered_values, axes=(0, 2, 1)
+            )
+            sampling_value_l_ = keras.ops.reshape(
+                permuted_gathered_values, (B_prime, hidden_dim, Q_dim, P_level)
+            )
+        else:
+            sampling_value_l_ = grid_sample(
+                data=value_l_,
+                grid=sampling_grid_l_,
+                align_corners=False,
+                height=height,
+                width=width,
+            )
+        sampling_value_list.append(sampling_value_l_)
+    # (batch_size, num_queries, num_heads, num_levels, num_points)
+    # -> (batch_size, num_heads, num_queries, num_levels, num_points)
+    # -> (batch_size, num_heads, 1, num_queries, num_levels*num_points)
+    _attention_weights = keras.ops.transpose(
```

Comment:
no underscore markers on local variables (here and everywhere). only for properties on a class, functions, methods, etc.